src/format/OpVaultReader.cpp

```diff
@@ -0,0 +1,498 @@
+#include "OpVaultReader.h"
+#include "OpData01.h"
+
+#include "core/Group.h"
+#include "core/Tools.h"
+#include "crypto/CryptoHash.h"
+#include "crypto/SymmetricCipher.h"
+
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QUuid>
+#include <QtGui/QtGui>
+#include <gcrypt.h>
+
+static void populateCategoryGroups(Group* rootGroup);
+
+OpVaultReader::~OpVaultReader()
+{
+    delete m_masterKey;
+    delete m_masterHmacKey;
+    delete m_overviewKey;
+    delete m_overviewHmacKey;
+}
+
+Database* OpVaultReader::readDatabase(QDir& opdataDir, const void* passPhrase, size_t passPhraseLen)
+{
+    auto result = new Database();
+    readDatabase(opdataDir, result, passPhrase, passPhraseLen);
+    return result;
+}
+
+void OpVaultReader::readDatabase(QDir& opVaultDir, Database* db, const void* passPhrase, size_t passPhraseLen)
+{
+    if (!opVaultDir.exists()) {
+        m_error = true;
+        m_errorStr = "Directory .opvault must exist";
```

Comment:
I don't think the class should keep state for error messages. They should parameters passed by reference. Also, all messages need to be translatable.